Use Salt for mach bootstrap

aneeshusa opened the pull request:
<!-- Please describe your changes on the following line: -->

This is currently WIP, but allows Salt for `mach bootstrap`. Not looking for review yet, just posting for visibility. You can run `./mach bootstrap` and Salt will run, letting you know what changes it would make for bootstrapping (doesn't actually run yet though).

Currently, this reads from saltfs using gitfs, meaning it always tracks the master branch. (Note that this is blocked on https://github.com/servo/saltfs/pull/577 landing; in the meantime, I've included a small workaround in this PR to pull from my updated saltfs branch, which will need to be removed.) In the future, the relevant Salt code may be merged in tree as part of Docker support for TC, and the bootstrapper should be updated to read from in tree.

Also, our Windows machines have not been Salted, so the existing Windows bootstrappers are retained. 

TODO:
- [x] Hook into existing bootstrapping code less hackily
- [x] Actually bootstrap instead of always using `test=True` mode (includes sudo)
- [x] Default to interactive mode (test first, then ask), with a force flag
- [x] Don't require running from the repository root directory
- [x] Make it easy to add support for other distros 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [x] These changes do not require tests because they should be verified manually

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14974)
<!-- Reviewable:end -->

aneeshusa committed:
Add a Salt bootstrapper

The Salt bootstrapper invokes Salt during `./mach bootstrap`
to install Servo's build dependencies.
It uses salt-call pinned to the same version of Salt as used in saltfs.

Currently, the implementation uses gitfs and reads directly from
the master branch of the saltfs repo;
in the future this should be changed when the relevant Salt states
are moved in-tree as part of Dockerization for TaskCluster.

We have not Salted our Windows machines, so the existing Windows
bootstrappers are retained. Currently this is only tested on
Ubuntu Trusty.

Salt uses various system python libraries,
including `python-apt` on Debian-based OSes to interact with apt.
`python-apt` does not seem to be installable via a requirements.txt
file, and the versions available on PyPI are far behind the versions
installed on actual Ubuntu machines.
Additionally, adding `python-apt` as an unconditional python dependency
would add bloat for users of other OSes, and lead to more churn
as additional OSes are supported.
However, as `python-apt` is already installed via apt on these machines,
we can allow Salt to instead use the module by using
`--system-site-packages` for the python virtualenv.
We also add the `-I` flag to `pip install` to ensure we have a local,
untouched copy of any other python packages used.
However, because this prints system-level Python packages in scope,
it slightly breaks isolation, so it is important to always pin
all dependencies in the requirements files.

highfive commented:
Heads up! This PR modifies the following files:
 * @wafflespeanut: python/requirements.txt, python/servo/bootstrapper/bootstrap.py, python/servo/bootstrapper/salt_config/minion, python/servo/bootstrapper/salt.py

aneeshusa commented:
This should work on Ubuntu Trusty now.

@Wafflespeanut @frewsxcv Python question: So far it seems necessary to use `--system-site-packages` when creating the virtualenv to allow Salt to work properly (see the commit message for details). Does this seem OK to you or should I make a separate virtualenv for Salt?

aneeshusa committed:
Remove unusable --android flag for `mach bootstrap`

`mach` can't do any bootstrapping for Android, so the flag is useless.

aneeshusa committed:
Create a util.py Python module for common functions

Extracting these functions helps avoid circular dependencies,
and make them easier to find/reuse.

aneeshusa committed:
Clean up and simplify existing `mach bootstrap`

- Default to interactive mode and remove the `--interactive` flag
- Use `--force` to skip interactivity
- Change MSVC dependency storage organization on disk: put each version
  into its own folder and directly refer to the versioned folders,
  providing immutability and making the installation list redundant
- Reuse `host_triple()` function to fix broken bootstrapper dispatching
- Simplify code:
  - Remove or inline many unused and redudant functions and variables
  - Prefer plain functions to classes
  - Consolidate into fewer files, remove unnecessary bootstrapper/ dir
- Improve Python style
- Sort dependency list

aneeshusa commented:
I've finished the remaining items on this, including a cleanup/overhaul of the existing bootstrapping. This is ready for review.

cc @UK992 since you wrote the original Windows bootstrapping, can you check that `mach bootstrap` still works on Windows?

aneeshusa commented:
Note this is still blocked on servo/saltfs#577 landing, however.

Wafflespeanut commented on python/servo/bootstrap.py:
```diff
@@ -0,0 +1,190 @@
+        proceed = raw_input(
+            'Proposed changes are above, proceed with bootstrap? [y/N]: '
+        )
+        if proceed not in ['y', 'yes', 'Y', 'Yes', 'YES']:
```
`if proceed.lower() not in ['y', 'yes']`? :smile:

Wafflespeanut commented:
This is great! :smile: 

> Does this seem OK to you or should I make a separate virtualenv for Salt?

This should be fine, since `pip install -I` is going to reinstall our local requirements which will shadow the globally available libraries anyway.

aneeshusa commented:
Also, I think this should work on macOS (using Homebrew) - if someone could check that I'll enable the Salt bootstrapper for macOS as well (instead of just Ubuntu 14.04).

UK992 commented:
Since both gnu and msvc use same python, detection should be relied on `host_triple`.
For msvc, maybe could `--force` flag delete `msvc-dependencies` folder and reinstall all packages?
Also salt should be excluded  in `requirements.txt` for Windows, because it's failed to create virtualenv.

UK992 commented:
Also `set()` on `WINDOWS_GNU` packages list should be removed, because it fails to execute command.

aneeshusa commented:
@UK992 
- I haven't changed the GNU/MSVC detection from the original. I'm not sure exactly how to use `host_triple`, can you show me how I should use it here?
- We shouldn't never need to delete the `msvc-dependencies` folder, since each package version is installed in its folder and never changed, so IMO a `--force` flag is unnecessary. If we do want that behavior it's different from what `--force` does for the other two, so I think we would need two flags at that point.
- Addressed your other two comments.

UK992 commented:
GNU/MSVC detection was broken for some time. Using `host_triple`:
```python
if "windows-gnu" in host_triple():
    bootstrapper = windows_gnu
elif "windows-msvc" in host_triple():
    bootstrapper = windows_msvc
```

aneeshusa committed:
Avoid hardcoded references to rustc in download()

Use the `desc` parameter instead to make the error messages
customized for the actual download.

Also use new-style format strings.

aneeshusa commented:
@UK992 OK thanks, I updated that. Can you give it a try and also run `./mach build` on MSVC after bootstrapping? I want to ensure the reorganizing of `msvc-dependencies` didn't break the `build_env` function.

UK992 commented:
@aneeshusa building works perfectly on MSVC.

aneeshusa commented:
servo/saltfs#577 merged and I've updated this, so this is no longer blocked.

jdm commented:
@metajack Review ping!

metajack commented:
@bors-servo r+

Looks good to me.

bors-servo commented:
:pushpin: Commit b4f5086 has been approved by `metajack`

<!-- @bors-servo r=metajack b4f508630510c1fc96f6f3245bfb29dd0e16dcf4 -->

bors-servo commented:
:hourglass: Testing commit b4f5086 with merge 1f76aa6...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4976), [arm32](http://build.servo.org/builders/arm32/builds/5145), [arm64](http://build.servo.org/builders/arm64/builds/5189), [linux-dev](http://build.servo.org/builders/linux-dev/builds/5290), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1964), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1974), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/5260), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/5300), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1962), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/2008), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/428), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/430)
